---
title: "Transcriptome Analysis"
author: "Nicholas Peterson"
date: "`r Sys.Date(5/9/2024)`"
output: pdf_document
---

# Load required packages (you might have to figure out how to install some of these first...)
```{r message=FALSE}
library(ballgown)
library(RColorBrewer)
library(genefilter)
library(dplyr)
library(devtools)
```


# What is this code doing? The code is creating a data frame with columns ids (containing "plank01", "plank02", "biofilm01", "biofilm02") and stage ("planktonic", "planktonic", "biofilm", "biofilm")
```{r}
pheno_data<-data.frame(ids = c("test01", "test02", "control01", "control02"),
                       stage = c("test", "test", "control", "control"))
```

# create Ballgown object and check transcript number
```{r}
library(ballgown)
```

```{r message=FALSE}
samples.c <- paste('ballgown', pheno_data$ids, sep = '/')
bg <- ballgown(samples = samples.c, meas='all', pData = pheno_data)
bg
```

install.packages("ballgown")


# <what is this code doing?> It's bringing back a previously saved R object to my current R session/environment. What data is in the bg_filt.RData file, not sure yet. 
```{r}
bg_filt = subset(bg,"rowVars(texpr(bg)) >1",genomesubset=TRUE)
bg_filt
```

# create a table of transcripts
```{r}
results_transcripts<- stattest(bg_filt, feature = "transcript", covariate = "stage",
getFC = TRUE, meas = "FPKM")
results_transcripts<-data.frame(geneNames=geneNames(bg_filt),
transcriptNames=transcriptNames(bg_filt), results_transcripts)
```


# Create a table of transcripts
```{r}
results_transcripts <- data.frame(geneNames = geneNames(bg_filt), transcriptNames = transcriptNames(bg_filt), results_transcripts)
```

# choose a transcript to examine more closely (this is a demo, you need to choose another)
```{r}
results_transcripts[results_transcripts$transcriptNames == "gene-EC110512_RS00015", ]
```
## what information are you given about this transcript? I think it's the statistics or measurements associated with our transcripts. The fold change, and significance (p-value, q-value), and other stuff

# <insert brief explanation of this code> It's filtering our transcript data,  by only taking data with a p-value less than .05.And the dim(sigdiff) command shows us the number of rows and columns in the significance/difference data frame.
```{r}
sigdiff <- results_transcripts %>% filter(pval<0.05)
dim(sigdiff)
```

# organize the table <by what metrics is the table being organized?> the table is being organized by p-value and absolute fold change.
```{r}
o = order(sigdiff[,"pval"], -abs(sigdiff[,"fc"]), decreasing=FALSE)
output = sigdiff[o,c("geneNames","transcriptNames", "id","fc","pval","qval")]
write.table(output, file="SigDiff.txt", sep="\t", row.names=FALSE, quote=FALSE)
head(output)
```

# load gene names
```{r}
bg_table = texpr(bg_filt, 'all')
bg_gene_names = unique(bg_table[, 9:10])
```

# pull out gene expression data and visualize
```{r}
gene_expression = as.data.frame(gexpr(bg_filt))
head(gene_expression)
```

# <what is this code doing? hint:compare the above output of head(gene_expression) to this output> All the code did was rename the the columns??
```{r}
colnames(gene_expression) <- c("test01", "test02", "control01", "control02")
head(gene_expression)
dim(gene_expression)
```

# load the transcript to gene table and determine the number of transcripts and unique genes
```{r}
transcript_gene_table = indexes(bg)$t2g
head(transcript_gene_table)
length(row.names(transcript_gene_table))
length(unique(transcript_gene_table[,"g_id"]))
```

# plot the number of transcripts per gene
```{r}
counts=table(transcript_gene_table[,"g_id"])
c_one = length(which(counts == 1))
c_more_than_one = length(which(counts > 1))
c_max = max(counts)
hist(counts, breaks=50, col="bisque4", xlab="Transcripts per gene",
main="Distribution of transcript count per gene")
legend_text = c(paste("Genes with one transcript =", c_one),
paste("Genes with more than one transcript =", c_more_than_one),
paste("Max transcripts for single gene = ", c_max))
legend("topright", legend_text, lty=NULL)
```

## <insert your interpretation of the above graph> Essentially all the genes only have one transcript, while very few have 2 transcripts, even less have 3 transcripts, and even less have 4 transcripts.

# create a plot of how similar the two replicates are for one another. We have two data sets...how can you modify this code in another chunk to create a plot of the other set?
```{r}
x = gene_expression[,"test01"]
y = gene_expression[,"test02"]
min_nonzero=1
plot(x=log2(x+min_nonzero), y=log2(y+min_nonzero), pch=16, col="purple", cex=0.25,
xlab="FPKM (test, Replicate 1)", ylab="FPKM (test, Replicate 2)",
main="Comparison of expression values for a pair of replicates")
abline(a=0,b=1, col = "hotpink")
rs=cor(x,y)^2
legend("topleft", paste("R squared = ", round(rs, digits=3), sep=""), lwd=1, col="black")
```

```{r}
x = gene_expression[,"control01"]
y = gene_expression[,"control02"]
min_nonzero=1
plot(x=log2(x+min_nonzero), y=log2(y+min_nonzero), pch=16, col="purple", cex=0.25,
xlab="FPKM (test, Replicate 1)", ylab="FPKM (test, Replicate 2)",
main="Comparison of expression values for a pair of replicates")
abline(a=0,b=1, col = "hotpink")
rs=cor(x,y)^2
legend("topleft", paste("R squared = ", round(rs, digits=3), sep=""), lwd=1, col="black")
```

## What does it mean if the two data sets are similar? Then the replicates are very similar.

# create plot of differential gene expression between the conditions
```{r}
results_genes = stattest(bg_filt, feature="gene", covariate="stage", getFC=TRUE, meas="FPKM")
results_genes = merge(results_genes,bg_gene_names,by.x=c("id"),by.y=c("gene_id"))
sig=which(results_genes$pval<0.7)
results_genes[,"de"] = log2(results_genes[,"fc"])
hist(results_genes[sig,"de"], breaks=50, col="seagreen",
xlab="log2(Fold change) Test vs Control",
main="Distribution of differential expression values")
abline(v=-2, col="black", lwd=2, lty=2)
abline(v=2, col="black", lwd=2, lty=2)
legend("topleft", "Fold-change > 4", lwd=2, lty=2)
```

## interpret the above figure: It shows the genes with statistically significant differences in expression. How different they are essentially visualized in a bar graph.

# Plot total gene expression highlighting differentially expressed genes
```{r}
gene_expression[,"test"]=apply(gene_expression[,c(1:2)], 1, mean)
gene_expression[,"control"]=apply(gene_expression[,c(3:4)], 1, mean)
x=log2(gene_expression[,"test"]+min_nonzero)
y=log2(gene_expression[,"control"]+min_nonzero)
plot(x=x, y=y, pch=16, cex=0.25, xlab="Test FPKM (log2)", ylab="Control FPKM (log2)",
main="Test vs Control FPKMs")
abline(a=0, b=1)
xsig=x[sig]
ysig=y[sig]
points(x=xsig, y=ysig, col="magenta", pch=16, cex=0.5)
legend("topleft", "Significant", col="magenta", pch=16)
```

# make a table of FPKM values
```{r}
fpkm = texpr(bg_filt,meas="FPKM")
```

# choose a gene to determine individual expression (pick a different number than I did)1234
```{r}
ballgown::transcriptNames(bg_filt)[1234]
ballgown::geneNames(bg_filt)[1234]
```
# transform to log2
```{r}
transformed_fpkm <- log2(fpkm[2, ] + 1)
```

# make sure values are properly coded as numbers
```{r}
numeric_stages <- as.numeric(factor(pheno_data$stage))

jittered_stages <- jitter(numeric_stages)
```

# plot expression of individual gene
```{r}
boxplot(transformed_fpkm ~ pheno_data$stage,
        main=paste(ballgown::geneNames(bg_filt)[3], ' : ', ballgown::transcriptNames(bg_filt)[3]),
        xlab="Stage",
        ylab="log2(FPKM+1)",
        col=c("lightblue", "salmon"),
        border="darkblue")

points(transformed_fpkm ~ jittered_stages, 
       pch=21, col="blue", bg="lightblue", cex=1.2)
```

## interpret the above figure:  It's showing how the expression of the "gene-EC110512_RS00015" 
15 gene varies between control and test. since the scale on the left is in log, it seems they are pretty similar but still different.

